### PR TITLE
Implement basic Composer targets

### PR DESCRIPTION
**Pull request by ezzatron**
Gotta start somewhere, right?

There's a ton of stuff to get your input on:

- Should it be `pkg-composer` or `pkg-php`? I went with the former, because I know that when I tackle JS, NPM vs Yarn will be a real thing I have to contend with. I also wasn't sure whether multiple "primary" Makefiles in a single `pkg-` repo is possible or advisable.
- I avoided the use of `composer update`. The intent is that this makefile should create a lockfile for you if none exists, but it shouldn't unexpectedly update dependencies for you. This is something that I learned from a similar Makefile for JS/Yarn. I found that I prefer to manage dependency updates explicitly outside of make targets, and it was more annoying when make "spookily" updated things for you.
- I couldn't find any prior art for an explicit `lint` target, or for the technique of using `tee` to produce command output logs as artifacts.
- There was a lot more complexity in some of the makefiles I was sourcing these targets from. They covered things like:
  - Installing Composer from source when not found in PATH
  - Auth credential management, including copying stuff from your home directory into `artifacts`, and setting `COMPOSER_HOME`. I think this was to facilitate doing `composer install` as a Docker build stage in a multi-stage build, when there are private Composer repositories involved.
- Some [`composer install`](https://getcomposer.org/doc/03-cli.md#install-i) options to consider:
  - `--prefer-dist` - "This can speed up installs substantially on build servers and other use cases where you typically do not run updates of the vendors."
  - `--no-dev` - For JS, I have some makefiles that automatically trim dev deps. It reduces how much data ends up in Docker images, for one thing.
  - `--no-suggest` - Suggestions are pretty annoying.
  - `--optimize-autoloader` - "This is recommended especially for production, but can take a bit of time to run so it is currently not done by default."
  - `--ignore-platform-reqs` - I think we can do without this now.
- Some [`composer validate`](https://getcomposer.org/doc/03-cli.md#validate) options to consider:
  - `--strict` - "Return a non-zero exit code for warnings as well as errors."
  - `--no-check-publish` - Maybe sometimes we don't want this?

Other commands to be considered for new targets:
- [`composer outdated`](https://getcomposer.org/doc/03-cli.md#outdated) - Pretty useful, and similar things exist for NPM/Yarn. Composer's defaults are dumb because they show you outdated *transitive* dependencies, which is useless and takes forever. The `--direct` and `--strict` options are probably sane defaults.

### CODE REVIEW

**Comment by jmalloc** [comment]
OMG yes, thank you.  I will look through this a little later.

**Comment by jmalloc** [comment]
> * Should it be `pkg-composer` or `pkg-php`?

I think in this case it should be `pkg-php`, because there's no other logical choice for PHP.  I suppose it could come about in the future, but I think at that point I'd want to decide whether there should be a `php` and `composer` and a `<whatever-new-dep-tool>`, or whether the `php` one should try to handle both.  

Obviously you have that decision to make for NodeJS, but I also don't think that all languages necessarily need to work the same; it just needs to make the most sense for the individual language... hopefully?

> * I avoided the use of `composer update`

Fair enough. This more closely matches how the Go makefiles work, too.  Can you remind me when our old makefiles ran composer update.  Was it when `.json` was newer than `.lock`?

> * I couldn't find any prior art for an explicit `lint` target, or for the technique of using `tee` to produce command output logs as artifacts.

So far any linting has been performed in `prepare`, I'm not sure if there's a reason to separate them (it certainly doesn't seem that way for Go, at least), but I'm open to it if there's need elsewhere and it keeps the "standard targets" uniform across languages.

> * Installing Composer from source when not found in PATH

Does it install it into `artifacts`? I've tended to move away from this kind of behavior in this org, and prefer just listing such things as a pre-requisite. I don't know that that necessarily needs to be an organisation-wide approach; if installing composer works well in practice I don't see any particular reason to stop.  

I do have something in mind for versioning binary dependencies for Go, which would involve installing them somewhere under the `.makefiles` tree.

> * Auth credential management

Yeah I remember this; it was for Docker, and I'm pretty sure it's *only* in the cwx makefiles, not jmalloc/rinq/icecave/etc.  I don't know how I feel about this one; on the surface it doesn't seem broadly application enough, or perhaps a bit hacky.


> * Some [`composer install`](https://getcomposer.org/doc/03-cli.md#install-i) options to consider:

* `--prefer-dist` - I think we used to do this in Archer.  FWIW we had to switch to `--prefer-source` (or nothing?) at cwx *before* Private Packagist because we needed SSH auth for GHE. I'm guessing this would work fine now.
* `--no-dev` - I remember when you implemented this logic. IIRC it was pretty involved (in the makefile) to switch back and forward between dev/no-dev, but I'm not sure if that was NPM specific problems. You've been using it for a while now and I like the idea of the smaller docker builds, for sure, so I'll leave this to your judgement.
* `--no-suggest` - Does this suppress the suggestions made when a package is archived? If not, definitely get rid of these.
> * `--optimize-autoloader` - Perhaps this could be done whenever `--no-dev` happens?
> * `--ignore-platform-reqs` - Cool.  I had been meaning to re-get-my-head-around this and the PHP version that we specify in the composer file under `config.platform`, I can't say I really remember why it was all necessary at this point.

> Some [`composer validate`](https://getcomposer.org/doc/03-cli.md#validate) options to consider:
* `--strict` - Makes sense
* `--no-check-publish` - This **probably** needs to be passed for applications, but not libraries.  IIRC we only make this distinction in any detectable way in NPM's package.json, but at cwx, and certainly for my own projects a repo is always an app or a library, and never both. There are other scenarios where we need to make this decision, at least at cwx, such as the fact that Go modules need to be tagged with a leading `v`, but we do not include a leading `v` for application builds (rollout, docker, etc). Could we perhaps either detect this somehow, or otherwise force the user to specify this in the makefile?


> * composer outdated

Can you think of any point when this could be run "automatically" for the user (perhaps `prepare`, if it's not too slow?)  If such a target were only present to "remind" the user to use `--direct` and `--strict` I'd probably vote to just leave it out.

This is almost verging on the stuff that we talked about briefly earlier in the week - scheduled CI builds that bump all dependencies to the latest, etc.

**Comment by jmalloc** [comment]
On, and in regards to `tee`, I have just been using a [`.touch` file](https://github.com/make-files/pkg-docker/blob/v1/Makefile#L38) instead.  I personally have never once looked through the log files.

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+.PHONY: prepare
+prepare:: artifacts/composer/validate.out
+
+################################################################################
+
+vendor: composer.lock
+	composer install
+
+composer.lock: composer.json
+	composer install
```
Comment (jmalloc) [inline Makefile]: I get why this is how it be; but this will trip `make` up somewhat.  `composer install` doesn't *actually* update `composer.lock`, so it will forever be considered out of date once it exists but `composer.json` is newer.

You could mark `composer.json` an "order only pre-requisite" (see bottom of this comment), which should mean that `make` will understand that it needs `composer.json` to build the lock file, but wont consider the lock file to be out of date just because `composer.json` is newer.

Should we be concerned about people forgetting to rebuild the lock file?  Should we check this in `prepare` and/or `ci`?
---

```Makefile
composer.lock : | composer.json
	composer install
```

File: Makefile [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+composer.json:
+	composer init --no-interaction
+
+artifacts/composer/validate.out: composer.json
+	@mkdir -p "$(@D)"
+	composer validate --no-check-publish | tee "$@
```
Comment (jmalloc) [inline Makefile]: Missing closing quote on argument to `tee`, but also see https://github.com/make-files/pkg-composer/pull/1#issuecomment-552075563

**Comment by ezzatron** [comment]
> I suppose it could come about in the future, but I think at that point I'd want to decide whether there should be a `php` and `composer` and a `<whatever-new-dep-tool>`, or whether the `php` one should try to handle both.

Can packages have multiple "main" makefiles at the moment? Just trying to think how I would approach the makefiles if I had to implement them today.

> Can you remind me when our old makefiles ran composer update. Was it when `.json` was newer than `.lock`?

Exactly that.

> So far any linting has been performed in `prepare`, I'm not sure if there's a reason to separate them (it certainly doesn't seem that way for Go, at least), but I'm open to it if there's need elsewhere and it keeps the "standard targets" uniform across languages.

I'm in the habit of running lint periodically mid-commit, and it might be annoying to always have to wait for tests too. One thing I use it for, for example, is pasting JSON into a JS file and letting lint "fix" it to be more readable (unquoting keys, turning double quotes in to single).

> Does it install it into `artifacts`? I've tended to move away from this kind of behavior in this org, and prefer just listing such things as a pre-requisite. I don't know that that necessarily needs to be an organisation-wide approach; if installing composer works well in practice I don't see any particular reason to stop.

Yeah, it did. Then it set a `COMPOSER` variable to point to that instead of `which composer`. I haven't re-implemented that here, I personally never used it, and I'm happy to let it go until it's needed.

> Yeah I remember this; it was for Docker, and I'm pretty sure it's _only_ in the cwx makefiles, not jmalloc/rinq/icecave/etc. I don't know how I feel about this one; on the surface it doesn't seem broadly application enough, or perhaps a bit hacky.

Well, even at CWX we ended up steering away from multi-stage builds, which was the use case for this anyway. I think it's best to leave it out for now 👍

> * `--no-suggest` - Does this suppress the suggestions made when a package is archived? If not, definitely get rid of these.

I think it just suppresses the suggestions on install: "Skips suggested packages in the output.".

> * Perhaps this could be done whenever `--no-dev` happens?

Yeah, that could be another good argument for having separate dev / non-dev `vendor` targets. Any thoughts on what to call those targets? They'd need to be used in dependency lists of other targets instead of just `vendor`.

> * I had been meaning to re-get-my-head-around this and the PHP version that we specify in the composer file under `config.platform`

Basically, the `config.platform` controls what ends up in your lock file, and `--ignore-platform-reqs` will install deps even if the "current" platform doesn't match their `php` or `ext-*` constraints.

> * `--strict` - Makes sense

Are you sure about that one? Are there any `composer validate` warnings that are stupid, or annoying to fix? I seem to remember that there were some weird ones.

> * Could we perhaps either detect this somehow, or otherwise force the user to specify this in the makefile?

Best candidate is probably https://getcomposer.org/doc/04-schema.md#type, which can be `library` or `project`.

> Can you think of any point when this could be run "automatically" for the user

Probably not.

File: Makefile (hunk 1 above)
Comment (ezzatron) [reply Makefile]: > Should we be concerned about people forgetting to rebuild the lock file? Should we check this in `prepare` and/or `ci`?

I think `composer validate` will catch it, because it has a `--no-check-lock` flag that's described as "Do not emit an error if composer.lock exists and is not up to date."

Order-only dep sounds like a good solution.

**Comment by ezzatron** [comment]
So I think I'll have a crack at doing the dev / non-dev `vendor` installs. Although there's probably a better way to do it than how it was done for JS at CWX.

File: Makefile (hunk 1 above)
Comment (jmalloc) [reply Makefile]: And we'd consider `composer validate` `lint`ing, and hence also run it via `prepare`?

**Comment by jmalloc** [comment]
> Can packages have multiple "main" makefiles at the moment? Just trying to think how I would approach the makefiles if I had to implement them today.

Yeah, I suppose they could - there's nothing in the makefiles themselves that means it is the "main" one. I've made sure, for example, that makefiles do not share variable names (unless they are defined in `lib-core`), and also been making more liberal use of the "double-colon" targets, so hopefully it's really just down to include-order as to how they interact.

That said, I'd also like to consider making each language package smart enough to handle the various different "mixins". Potentially, this could be accomplished by `include`ing more specific makefiles when certain files are present in the repository. For example, having `pkg-php/Makefile` include `composer.mk` if `composer.json` is present, and `phpunit.mk` if `phpunit.xml[.dist]` is present, etc.

> I'm in the habit of running lint periodically mid-commit,

:+1: ok, let's retain a `lint` target

> * `--strict` - Makes sense
> Are you sure about that one?

No, not 100%, it's been so long since I've done daily PHP development! Still, I figure it's better to include it initially and relax it later than the other way around. When we migrate projects to use these makefiles, if we discover a lot of unwarranted failures we can very easily remove it.

> Yeah, that could be another good argument for having separate dev / non-dev `vendor` targets. Any thoughts on what to call those targets? They'd need to be used in dependency lists of other targets instead of just `vendor`.

Presumably these will be "touch" files or similar that reflect the state of the actual `vendor`? If so, I think `artifacts/touch/php/vendor.[no-]dev` is fine.

> Best candidate is probably https://getcomposer.org/doc/04-schema.md#type, which can be `library` or `project`.

If there's an easy way to pull that out, that seems like a good idea to me. 

---

There's a lot of things going on here and I fear this thread will get too hard to follow if we keep replying like this. I think the most pressing thing to work out is probably the first point about the naming of this package and how "mixin" type behavior will be achieved, and anything else can probably be put into a separate issue?

**Comment by ezzatron** [comment]
> There's a lot of things going on here and I fear this thread will get too hard to follow if we keep replying like this. I think the most pressing thing to work out is probably the first point about the naming of this package and how "mixin" type behavior will be achieved, and anything else can probably be put into a separate issue?

Sure. Happy to switch back to `pkg-php`. I like the idea of conditional mixin includes dependant on files existing. Although for dependency managers, it might be best to fall back to a default. That way, we can retain this type behaviour, which is quite nice for starting new projects:

```Makefile
composer.json:
	composer init --no-interaction
```

**Comment by jmalloc** [comment]
> Although for dependency managers, it might be best to fall back to a default.

Yep this sounds good to me, generally we'll probably have some idea of what we consider to be "best practice" at any given time, so that should be the default. This could possibly happen with test frameworks too? Bring on Recoil+Kahlan+Phony?

File: Makefile (hunk 1 above)
Comment (ezzatron) [reply Makefile]: Well, in the implementation I had here, `composer validate` was an explicit part of `prepare`. I don't know if I'd consider it linting, but we can discuss that on another issue.

**Comment by ezzatron** [comment]
Okay, I think it's pretty close as a starting point. I broke Composer stuff out into its own file, that is always included for now.

I will tackle the conditional use of `--no-check-publish` in a another PR.
